core/src/main/java/org/elasticsearch/index/snapshots/blobstore/BlobStoreIndexShardRepository.java

```diff
@@ -359,38 +359,32 @@ public BlobStoreIndexShardSnapshot loadSnapshot() {
          */
         protected void finalize(List<SnapshotFiles> snapshots, int fileListGeneration, Map<String, BlobMetaData> blobs) {
             BlobStoreIndexShardSnapshots newSnapshots = new BlobStoreIndexShardSnapshots(snapshots);
-            List<String> blobsToDelete = new ArrayList<>();
             // delete old index files first
-            for (String blobName : blobs.keySet()) {
-                // delete old file lists
+            blobs.keySet().forEach(blobName -> {
                 if (indexShardSnapshotsFormat.isTempBlobName(blobName) || blobName.startsWith(SNAPSHOT_INDEX_PREFIX)) {
-                    blobsToDelete.add(blobName);
+                    try {
+                        blobContainer.deleteBlob(blobName);
+                    } catch (IOException e) {
+                        // We cannot delete index file - this is fatal, we cannot continue, otherwise we might end up
+                        // with references to non-existing files
+                        throw new IndexShardSnapshotFailedException(shardId, "error deleting index files during cleanup", e);
```

Comment:
Since we know the individual file here, we could give a more descriptive error message by saying "error deleting index file [" + blobName + "] during cleanup"